apis/v1/gatewayclass_types.go

```diff
@@ -278,6 +279,10 @@ type GatewayClassList struct {
 	Items           []GatewayClass `json:"items"`
 }
 
-// SupportedFeature is used to describe distinct features that are covered by
+// FeatureName is used to describe distinct features that are covered by
 // conformance tests.
-type SupportedFeature string
+type FeatureName string
+
+type SupportedFeature struct {
+	Name FeatureName `json:"name"`
+}
```

Comment:
speaking on behalf of Envoy Gateway, where we enabled this feature by default, our users have not found just the `name` to be very useful. As mentioned earlier an additional optional field like `Message` / `Description` / `Info` can inform the user, specific caveats that may be implementation specific that have be intentionally kept open in the API (where `SHOULD` and `SHOULD NOT` are used for e.g. in https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.HTTPRequestRedirectFilter)

Reply:
I think the rationale for `name` is that it leaves room to add additional fields in the future, but we couldn't quite agree on what those should look like yet. I do generally agree that a message field could be helpful here, but I think we'd want to agree on formatting, when/how it should be populated, etc, before actually adding it to the API.

Reply:
that rationale makes sense @robscott, a follow question is, are there any implementations that would like to implement this API in this current form ?

Reply:
IMO, this iteration of the API is primarily about automation. From the manual form of `kubectl get gatewayclass -oyaml | grep Redirect` to the more ideal goal where we have tooling like gwctl that can automatically warn if a user is trying to use an unsupported feature. Presumably in that case gwctl would also warn if a GatewayClass had not populated supported features.

I think what you're describing is primarily useful for someone that is trying to look at the full GatewayClass status and understand the nuance of what's supported, which I think is helpful, but not quite as high priority as the UX I described above.

Reply:
This API is the current form will benefit tooling  like gwctl, but imo is not useful enough for end users who do not use tools like gwctl, which imo should drive the design of this API.

Reply:
I don't think user documentation belongs in CRD Status at all. IMO this should be removed from the API and documented in implementations websites.

Reply:
> I don't think user documentation belongs in CRD Status at all. IMO this should be removed from the API and documented in implementations websites.

I am mostly plus one to this, except from the supported features/profiles which could be really useful and improve the UX(through warning and automations) for an api where implementations not necessarily support all features.

Reply:
> I don't think user documentation belongs in CRD Status at all. IMO this should be removed from the API and documented in implementations websites.

By "this", @howardjohn, do you mean SupportedFeatures in general, or the idea of a `message` field on each?

Personally, I'd like to see us do a few things once we get the current change in:
* Define all the `SupportedFeatures` in Gateway API code, with a standard description of each. Currently the only definition is in the test suite, but I think we may need to promote this into a separate package somewhere so that both tooling and the conformance suite can use it. Conceivably, we could also include support state and "lastTransitionVersion" or something to record the last time it changed too.
* Have a generated page on the website that lists all the supported features and their associated fields.
* Ensure that any Gateway API guides on the project page list which features are required for that guide, and encourage guide and documentation authors to list the required features on their guides as well.

If we do the above, a lot of the reason for having this information inlined in the GatewayClass status goes away, because the `SupportedFeature` name allows users or code to look up any associated data in the canonical location.

Having some form of `SupportedFeature` in GatewayClass status, along with a canonical list of `SupportedFeature` names, allows us to move towards having the conformance suite determine which tests to run by looking at the GatewayClass as well, which makes it theoretically possible to run the vanilla conformance test suite against _any_ Gateway API implementation, similarly to how you can run upstream Kubernetes conformance against _any_ cluster and get a reasonable result.

I think that it's still worth keeping this list as a list of named subobjects, even if we only have `name` for now, because of the expandability it gives us later - for the cost of a bunch of superfluous (for now) `name: ` strings.

Reply:
I'm +1 to the idea proposed here and how it's implemented (using a slice of objects with a single `name` field). I initially may have had some second thoughts around the need to have `SupportedFeatures` but have gotten around that after reading all the discussions :)

I like the vision that @youngnick has portrayed around the overlap with Conformance Suite / Profiles and how eventually this may serve as an input to generating the Conformance reports.

For this to be really useful with tooling like gwctl, we definitely should work towards a follow up for this change to come with with that canonical list of SupportedFeatures, or, I'm assuming we intend to use https://github.com/kubernetes-sigs/gateway-api/blob/d400a8b69ef19578d55896852b75d7d5e6f9a30c/pkg/features/features.go as the set of possible values while the field is in experimental.

Reply:
+1 on having this map of struct with just the `name` field for the time being. I agree that putting documentation or link to docs in the resource status is something we should try to avoid and lean toward creating a proper documentation in the canonical place instead.

I +1 the plan stated above by Nick, with just a small nit: we already have all the features in a separate [package](https://github.com/kubernetes-sigs/gateway-api/blob/89996b546f34a880b7dcf9ae101926d5b0463794/pkg/features), but we need to write appropriate go docs for all the features, as currently, they are mostly in the form of `This Option indicates support for feature X`, which is not useful at all for users.

> Conceivably, we could also include support state and "lastTransitionVersion" or something to record the last time it changed too.

Big +1 . Just listing the features is not enough. The addition of some metadata on the features is something fundamental in my opinion, to track the feature lifecycle and its implementation state.